packages/typespec-client-generator-core/src/decorators.ts

```diff
@@ -113,20 +124,53 @@ function setScopedDecoratorData(
     return;
   }
 
-  // if scope specified, create or overwrite with the new value
-  const splitScopes = scope.split(",").map((s) => s.trim());
-  const targetEntry = context.program.stateMap(key).get(target);
-
-  // if target doesn't exist in decorator map, create a new entry
-  if (!targetEntry) {
-    const newObject = Object.fromEntries(splitScopes.map((scope) => [scope, value]));
+  const negationScopes = getNegationScopes(scope);
+  if (negationScopes !== undefined && negationScopes.length > 0) {
+    const newObject = Object.fromEntries([AllScopes].map((scope) => [scope, value]));
+    newObject[negationScopesKey] = negationScopes;
     context.program.stateMap(key).set(target, newObject);
     return;
+  } else {
+    const splitScopes = scope.split(",").map((s) => s.trim());
+
+    // if negation scope is combined with normal scope, report error
+    if (splitScopes.length > 1 && splitScopes.some((s) => s.startsWith("!"))) {
+      reportDiagnostic(context.program, {
+        code: "invalid-negation-scope",
+        target: context.decoratorTarget,
+      });
+      return;
+    }
+
+    // if scope specified, create or overwrite with the new value
+    const targetEntry = context.program.stateMap(key).get(target);
+
+    // if target doesn't exist in decorator map, create a new entry
+    if (!targetEntry) {
+      const newObject = Object.fromEntries(splitScopes.map((scope) => [scope, value]));
+      context.program.stateMap(key).set(target, newObject);
+      return;
+    }
+
+    // if target exists, overwrite existed value
+    const newObject = Object.fromEntries(splitScopes.map((scope) => [scope, value]));
+    context.program.stateMap(key).set(target, { ...targetEntry, ...newObject });
```

Comment:
the main problem for the algorithm is it seems not support incremented set. for example:
```
@@doc(A, "else", "!java");
@@doc(A, "java", "java");
```

also, could you add some tests for that?

Reply:
added